---
title: "OpenViRome_Walkthrough"
author: "Aiden Hiller"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{OpenViRome_Walkthrough}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup, warning=FALSE, message=FALSE}
library(OpenViRome)
```

## Basic Overview

OpenViRome is an R package designed to generate 'virome' style reports of Serratus
data given a taxonomic group of interest. The package's primary goal is to
provide easy access to Serratus data in a meaningful framework for the purpose of
virus discovery. OpenViRome aims to take as input any arbitrary grouping of SRA data
(runs, taxonomic groups, etc.) and provide a number of visualizations that can
be used to identify novel viruses within that data of high priority. Visulizations
within the package can broadly be grouped into two categories:
1. Objective Plots. Such plots are not likely to be easily interpretable as 
metagenomic data is innately noisy, but rather they plainly lay out the data 
without bias. These plots are meant to answer:
 - What viruses are present in this taxon across the SRA?
 - How are the viruses distributed across species within the taxon?
 - How is viral diversity different between samples?
 
2. Opinionated Plots. These plots are meant to elucidate more of a narrative of
which viruses are high priority for characterization, based on features I have
deemed to be important (e.g. high coverage and low GenBank identity).

The focus of OpenViRome is to generate meaningful and interpretable plots of viral
diversity for streamlined analysis, not the generation of new data. It is meant
to be used in tandem with palmid, which provides a contrasting 'virus-first'
approach.

For this walkthrough, we will be focusing on characterizing the virome of the
parasitic nematode superfamily, Tylenchoidea. This group includes the genera
Globodera, Heterodera, and Meloidogyne, which parasitize a wide range of 
agricultural crops and are responsible for billions of dollars in crop loss
annually.

In this instance, we are interested in understanding Tylenchoidea viral diversity
with the goal of identifying potential biocontrol agents due to the high agricultural
relevance. However, this is not the only use case for OpenViRome.

================================================================================
## Part 1: Setup

```{r warning=FALSE, message=FALSE}
library(OpenViRome)

con <- palmid::SerratusConnect() # establish a connection to the Serratus SQL
#TylenchoideaVirome <- getVirome(tax="Tylenchoidea", con=con)

# Note that you can also use the 'getVirome' function to query SRA runs, i.e.
# getVirome(sra=c("SRR1234567", "SRR7654321"), con=con)
```
This cell may take a while to run, as it is querying the Serratus SQL for all
Tylenchoidea data. Larger taxonomic groups will take longer to query. 
Alternatively, you can load the included Tylenchoidea virome object.

```{r}
data("TylenchoideaVirome")
```


## Part 2: Objective Plots
Currently, the objective plots provided are:
 - Phylum abundance (piechart)
 - Sankey diagram of sOTUs across bioprojects
 - Barchart of virus-positive runs for each source species
 - Alpha diversity metrics for each biosample
 - Support for beta diversity (ordination) using the ampvis2 package.
 
 
```{r}
# Overview stats
getViromeSummary(TylenchoideaVirome)
```
 
```{r}
# Phylum abundance
plotViromePie(TylenchoideaVirome)
```

Here we'll get a warning that there are too many phyla to be discernible
by color. So we can instead plot a subset:

```{r}
plotViromePie(TylenchoideaVirome, phylumFilter=c("Negarnaviricota", 
                                                  "Lenarviricota"))
# Note that slices are now proportional to the filtered phyla, not the total.
```

We can also customize the colors:
```{r}
plotViromePie(TylenchoideaVirome, phylumFilter=c("Negarnaviricota", 
                                                  "Lenarviricota"),
               colors=c("red", "blue"))
```

And do the same for the sankey plot:
```{r}
drawVirusSankey(TylenchoideaVirome, phylumFilter=c("Negarnaviricota", 
                                                  "Lenarviricota"),
               colors=c("red", "blue"))
```

We can also plot the virus-positive runs for each source species. This might
give us a sense of whether a species might have a chronic/endogenous infection,
in addition to what kind of data is available for this taxon in the SRA.
```{r}
# First have to calculate 'host burden'
vbDF <- getHostViralBurden(TylenchoideaVirome)
# Then plot
plotVirusPositive(vbDF)
```

OpenViRome also supports plotting alpha and beta diversity of the virome. Here,
I treat each biosample as a 'community' and normalized read count as the number
of observations for each sOTU.
```{r}
# Alpha diversity
plotAlphaDiversity(TylenchoideaVirome, mode = "shannon")

# shannon diversity is plotted by default, but you can also plot simpson with 
# the 'mode' argument
```

We can also just calculate diversity metrics individually for the virome:
```{r}
# Shannon diversity
getAlphaDiversity(TylenchoideaVirome, mode = "shannon")

# Simpson diversity
getAlphaDiversity(TylenchoideaVirome, mode = "simpson")

# Also supports 'evenness' and 'richness' with the mode argument, and individual
# biosamples with the 'bioSample' argument.

```

Finally, OpenViRome provides support for transforming a virome object to an
ampvis2 object. Read counts are already normalized based on SRA run total reads.
```{r}
d <- viromeToOTUTable(TylenchoideaVirome)
m <- viromeToMetadataTable(TylenchoideaVirome) # metadata is optional

ampvis2::amp_load(otutable = d, metadata = m)
```

```{r}
# Can also just plot bray-curtis distance from within OpenViRome
plotBetaDiversity(TylenchoideaVirome)
```
This plot uses PCoA with bray-curtis and no transformation. For more options,
see the ampvis2 package [documentation](https://kasperskytte.github.io/ampvis2/articles/ampvis2.html)

## Part 3: Opinionated Plots

Currently OpenViRome has two opinionated plots that prioritize high divergence,
high coverage, and common SRA prevalence viruses.

```{r}
require(ComplexHeatmap)
# Clustered heatmap plotting viral phylum (of each sOTU) against SRA runs in
# which they appear
viroMap(TylenchoideaVirome, minCov=100) # the minCov argument filters out sOTUs
# with cov < minCov. Often the plot will be too large to be useful without this
# filter due to the high number of low coverage sOTUs.

# Unfortunately, this plot does not display properly in a knitted markdown or in
# the Rstudio viewer. It is best viewed in its own window.
```

```{r}
# And the scatterplot:

palmPrevalence(TylenchoideaVirome)

# Here each sOTU is represented as a point, with the x-axis being the number of
# SRA runs in which it appears, and the y-axis being the number of bio_samples.
# The size of the point is proportional to the mean number of normalized reads
# for that sOTU across all runs within the virome within which it appears.

# The color of each point is determined the sOTUs divergence from the closest
# alignment in GenBank.
```

## Part 4: Shiny App

OpenViRome was mostly developed to provide a backend for the Shiny interface,
which provides virome summary reports for any taxonomic group or set of SRA 
runs without requiring any code from the user. 

You can run the shiny app locally with the viromeDashboard function:
```{r}
#viromeDashboard()
```
